File: 347-top-k-frequent-elements/top-k-frequent-elements.py
from collections import Counter
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:

        count = Counter(nums)

        min_heap = []

        for key, value in count.items():
            heappush(min_heap, (value, key))

            while len(min_heap) > k:
                heappop(min_heap)

        # print("heap -> ", min_heap)
        result = []
        for key, value in min_heap:
            # print("values -> ", value)
            result.append(value)
        
        return result

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        Map<Integer, Integer> freqMap = new HashMap<>();

        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>((a,b) -> (a.getValue() - b.getValue()));

        for (Map.Entry<Integer, Integer> me : freqMap.entrySet()){
            
            pq.add(me);
            if (pq.size() > k) {
                pq.poll();
            }
        }
        
        int[] result = new int[k];
        int i = 0;
        while (!pq.isEmpty()) {
            result[i++] = pq.poll().getKey();
        }
        return result;
    }
}